# Black Friday (EDA) and Feature Engineering

### Cleaning and Preparing the data for model training

**Data set :** *https://www.kaggle.com/datasets/sdolezel/black-friday*

# Problem Statement
A retail company "ABC Private limied" want to understand the purchase behaviour (specifiacally,purchase amount)against various products of different categories.They have shared purchase summary of various cudtomerd for selected high volume products from last month.The dataset also contains customer demographics(age,gender,maritial  status,city type,stay in current city),product details (product id and product category) and tota purchase amount from last month.

Now,they want to build a model to predict the purchase amount of customer against product which will help them to create personalized offer for customers against different products.

In [9]:
# importing libaries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
%matplotlib inline
warnings.filterwarnings('ignore')

In [11]:
# importing dataset
df_train=pd.read_csv('train.csv')
df_train.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN      8370  
1                 6.0                14.0     15200  
2                 NaN                 NaN      1422  
3                14.0                 NaN      1057  
4                 NaN                 NaN      7969

In [13]:
# importing the test data
df_test=pd.read_csv('test.csv')
df_test.head()

User_ID Product_ID Gender    Age  Occupation City_Category  \
0  1000004  P00128942      M  46-50           7             B   
1  1000009  P00113442      M  26-35          17             C   
2  1000010  P00288442      F  36-45           1             B   
3  1000010  P00145342      F  36-45           1             B   
4  1000011  P00053842      F  26-35           1             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               1                   1   
1                          0               0                   3   
2                         4+               1                   5   
3                         4+               1                   4   
4                          1               0                   4   

   Product_Category_2  Product_Category_3  
0                11.0                 NaN  
1                 5.0                 NaN  
2                14.0                 NaN  
3                 9.0                 NaN  
4                 5.0                12.0

# Why to merge Test and Train data?
When you perform preprocessing steps such as encoding categorical variables, scaling features, or imputing missing values, it's crucial to apply the same transformations to both training and test sets. By merging them before preprocessing, you ensure that the same transformations are applied consistently.

In [15]:
# merging both train and test data
df=df_train.append(df_test)
df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN    8370.0  
1                 6.0                14.0   15200.0  
2                 NaN                 NaN    1422.0  
3                14.0                 NaN    1057.0  
4                 NaN                 NaN    7969.0

In [17]:
# Basic Code
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 783667 entries, 0 to 233598
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     783667 non-null  int64  
 1   Product_ID                  783667 non-null  object 
 2   Gender                      783667 non-null  object 
 3   Age                         783667 non-null  object 
 4   Occupation                  783667 non-null  int64  
 5   City_Category               783667 non-null  object 
 6   Stay_In_Current_City_Years  783667 non-null  object 
 7   Marital_Status              783667 non-null  int64  
 8   Product_Category_1          783667 non-null  int64  
 9   Product_Category_2          537685 non-null  float64
 10  Product_Category_3          237858 non-null  float64
 11  Purchase                    550068 non-null  float64
dtypes: float64(3), int64(4), object(5)
memory usage: 77.7+ MB


In [19]:
df.describe()

User_ID     Occupation  Marital_Status  Product_Category_1  \
count  7.836670e+05  783667.000000   783667.000000       783667.000000   
mean   1.003029e+06       8.079300        0.409777            5.366196   
std    1.727267e+03       6.522206        0.491793            3.878160   
min    1.000001e+06       0.000000        0.000000            1.000000   
25%    1.001519e+06       2.000000        0.000000            1.000000   
50%    1.003075e+06       7.000000        0.000000            5.000000   
75%    1.004478e+06      14.000000        1.000000            8.000000   
max    1.006040e+06      20.000000        1.000000           20.000000   

       Product_Category_2  Product_Category_3       Purchase              B  \
count       537685.000000       237858.000000  550068.000000  783667.000000   
mean             9.844506           12.668605    9263.968713       0.420764   
std              5.089093            4.125510    5023.065394       0.493682   
min              2.000000            3.000000      12.000000       0.000000   
25%              5.000000            9.000000    5823.000000       0.000000   
50%              9.000000           14.000000    8047.000000       0.000000   
75%             15.000000           16.000000   12054.000000       1.000000   
max             18.000000           18.000000   23961.000000       1.000000   

                   C  
count  783667.000000  
mean        0.310954  
std         0.462884  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [21]:
df.drop(['User_ID'],axis=1,inplace=True)

In [23]:
df

Product_ID Gender    Age  Occupation Stay_In_Current_City_Years  \
0       P00069042      F   0-17          10                          2   
1       P00248942      F   0-17          10                          2   
2       P00087842      F   0-17          10                          2   
3       P00085442      F   0-17          10                          2   
4       P00285442      M    55+          16                         4+   
...           ...    ...    ...         ...                        ...   
233594  P00118942      F  26-35          15                         4+   
233595  P00254642      F  26-35          15                         4+   
233596  P00031842      F  26-35          15                         4+   
233597  P00124742      F  46-50           1                         4+   
233598  P00316642      F  46-50           0                         4+   

        Marital_Status  Product_Category_1  Product_Category_2  \
0                    0                   3                 NaN   
1                    0                   1                 6.0   
2                    0                  12                 NaN   
3                    0                  12                14.0   
4                    0                   8                 NaN   
...                ...                 ...                 ...   
233594               1                   8                 NaN   
233595               1                   5                 8.0   
233596               1                   1                 5.0   
233597               0                  10                16.0   
233598               1                   4                 5.0   

        Product_Category_3  Purchase  B  C  
0                      NaN    8370.0  0  0  
1                     14.0   15200.0  0  0  
2                      NaN    1422.0  0  0  
3                      NaN    1057.0  0  0  
4                      NaN    7969.0  0  1  
...                    ...       ... .. ..  
233594                 NaN       NaN  1  0  
233595                 NaN       NaN  1  0  
233596                12.0       NaN  1  0  
233597                 NaN       NaN  0  1  
233598                 NaN       NaN  1  0  

[783667 rows x 12 columns]

In [25]:
# converting categorical variables into numerical variables(Gender)
df['Gender']=df['Gender'].map({'F':0,'M':1})
df.head()

Product_ID  Gender   Age  Occupation Stay_In_Current_City_Years  \
0  P00069042       0  0-17          10                          2   
1  P00248942       0  0-17          10                          2   
2  P00087842       0  0-17          10                          2   
3  P00085442       0  0-17          10                          2   
4  P00285442       1   55+          16                         4+   

   Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3  \
0               0                   3                 NaN                 NaN   
1               0                   1                 6.0                14.0   
2               0                  12                 NaN                 NaN   
3               0                  12                14.0                 NaN   
4               0                   8                 NaN                 NaN   

   Purchase  B  C  
0    8370.0  0  0  
1   15200.0  0  0  
2    1422.0  0  0  
3    1057.0  0  0  
4    7969.0  0  1

In [27]:
# # converting categorical variables into numerical variables(Age)
df['Age'].unique()

array(['0-17', '55+', '26-35', '46-50', '51-55', '36-45', '18-25'],
      dtype=object)

In [29]:
#pd.get_dummies(df['Age'],drop_first=True)
df['Age']=df['Age'].map({'0-17':1,'18-25':2,'26-35':3,'36-45':4, '46-50':5,'51-55':6,'55+':7})
df.head()

Product_ID  Gender  Age  Occupation Stay_In_Current_City_Years  \
0  P00069042       0    1          10                          2   
1  P00248942       0    1          10                          2   
2  P00087842       0    1          10                          2   
3  P00085442       0    1          10                          2   
4  P00285442       1    7          16                         4+   

   Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3  \
0               0                   3                 NaN                 NaN   
1               0                   1                 6.0                14.0   
2               0                  12                 NaN                 NaN   
3               0                  12                14.0                 NaN   
4               0                   8                 NaN                 NaN   

   Purchase  B  C  
0    8370.0  0  0  
1   15200.0  0  0  
2    1422.0  0  0  
3    1057.0  0  0  
4    7969.0  0  1

In [31]:
#converting categorical variables into numerical variables('City_Category')
df_city=pd.get_dummies(df['City_Category'],drop_first=True)
df_city.head()

KeyError: 'City_Category'

In [ ]:
# updating in data set
df=pd.concat([df,df_city],axis=1)
df.head()

In [ ]:
#drop city category feature
df.drop('City_Category',axis=1,inplace=True)

In [ ]:
df.head()

In [ ]:
# Finding Missing values
df.isnull().sum()

In [ ]:
# Focusing on replacing missing values ['Product_Category_2']
df['Product_Category_2'].unique()

In [ ]:
df['Product_Category_2'].value_counts()

In [ ]:
df['Product_Category_2'].mode()[0]

In [ ]:
# Best way to fix missing value with mode
df['Product_Category_2']=df['Product_Category_2'].fillna(df['Product_Category_2'].mode()[0])

In [ ]:
df['Product_Category_2'].isnull().sum()

In [ ]:
# Focusing on replacing missing values ['Product_Category_3']
df['Product_Category_3'].unique()

In [ ]:
df['Product_Category_3'].value_counts()

In [ ]:
# Best way to fix missing value with mode
df['Product_Category_3']=df['Product_Category_3'].fillna(df['Product_Category_3'].mode()[0])

In [ ]:
df.head()

In [ ]:
df.shape

In [ ]:
df['Stay_In_Current_City_Years'].unique()

In [ ]:
df['Stay_In_Current_City_Years']=df['Stay_In_Current_City_Years'].str.replace('+','')

In [ ]:
df.head()

In [ ]:
df.info()

In [ ]:
# convert object into integers
df['Stay_In_Current_City_Years']=df['Stay_In_Current_City_Years'].astype(int)
df.info()

In [ ]:
df['B']=df['B'].astype(int)
df['C']=df['C'].astype(int)

In [ ]:
df.info()

# Visualisation

In [ ]:
correlation_matrix = df.corr()
plt.figure(figsize=(20, 12))
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm')
plt.title('Correlation Heatmap')
plt.show()

In [ ]:
# visualisation
sns.barplot(x='Age', y='Purchase', hue='Gender', data=df)

# Observation(Age group analysis)
The bar plot shows consistent purchase amounts around 10,000 across all age groups. Males generally spend slightly more than females , especially in the younger ('0-17' and '18-25') and older ('51-55') age groups. Overall, gender differences in purchase amounts are minimal.

In [ ]:
# Visualization of purchase with occupation
sns.barplot(x='Occupation', y='Purchase', hue='Gender', data=df)

# Observation(Gender group analysis with respect to occupation)
The bar plot shows that purchase amounts are consistently close to 10,000 across all occupations. Males (1) generally spend slightly more than females (0) across most occupations, with minimal gender differences. Occupation 17 exhibits the highest purchase amounts, particularly for males.

In [ ]:
# Visualization of purchase with product category 1
sns.barplot(x='Product_Category_1',y='Purchase',hue='Gender', data=df)

# Observation:

1. The graph compares purchase amounts across 20 different product categories.
2. Both genders (labeled '0' for female and '1' for male) show similar purchasing behavior in most categories.
3. Purchase of 20000 is the highest purchase in product category 1

In [ ]:
sns.barplot(x='Product_Category_2',y='Purchase',hue='Gender', data=df)

# Observation:
1. The chart compares purchases across 20 product categories.
2. Both genders exhibit similar purchasing behavior in most categories.
3. However, certain categories show higher purchases by male over the female.
4. Purchase of 16000 is the highest purchase in product category 2

In [ ]:
sns.barplot(x='Product_Category_3',y='Purchase',hue='Gender', data=df)

# Observation:
1. The chart compares purchases across 20 product categories.
2. Both genders exhibit similar purchasing behavior in most categories.
3. However, certain categories show higher purchases by male over the female.
4. Purchase of 14000 is the highest purchase in product category 3

# Feature Scaling

In [ ]:
#feature scaling
df_test=df[df['Purchase'].isnull()]

In [ ]:
df_train=df[~df['Purchase'].isnull()]

In [ ]:
X=df_train.drop('Purchase',axis=1)

In [ ]:
X.head()

In [ ]:
X.shape

In [ ]:
y.shape

In [ ]:
y=df_train['Purchase']

In [ ]:
y

In [ ]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [ ]:
X_train.drop('Product_ID',axis=1,inplace=True)
X_test.drop('Product_ID',axis=1,inplace=True)

In [ ]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X_train=sc.fit_transform(X_train)
X_test=sc.transform(X_test)

In [ ]:
from sklearn.linear_model import LinearRegression

In [ ]:
model = LinearRegression()
model.fit(X_train, y_train)
w=model.coef_
b=model.intercept_
print(w)
print(b)

In [ ]:
predictions=model.predict(X_test)
predictions

In [ ]:
from sklearn.ensemble import RandomForestRegressor

In [ ]:
RF = RandomForestRegressor(n_estimators=10,max_depth=7,n_jobs=-1,random_state=42)
RF.fit(X_train,y_train)

In [ ]:
y_pred = RF.predict(X_test)
y_pred

In [ ]:
y_test

In [ ]:
!pip install xgboost

In [ ]:
from xgboost.sklearn import XGBRegressor
from sklearn.metrics import r2_score, mean_squared_error,mean_absolute_error

In [ ]:
MAE = mean_absolute_error(y_test,y_pred)
MAE

In [ ]:
r2_score(y_test, y_pred)

In [ ]:
print(mean_squared_error(y_test, y_pred))

In [ ]:
print(np.sqrt(mean_squared_error(y_test, y_pred)))

In [ ]:
y.mean()

In [ ]:
y_test.mean()

In [ ]:
xgb_reg=XGBRegressor(n_estimators=480, learning_rate=0.2, max_depth=8)
xgb_reg.fit(X_train,y_train)

In [ ]:
xgb_y_pred=xgb_reg.predict(X_test)

In [ ]:
print("RMSE:",np.sqrt(mean_squared_error(y_test, xgb_y_pred)))
print("y_mean",(y.mean()))
print("R2 Score:",round(r2_score(y_test,xgb_y_pred)*100,2))

### **Summary:**
**RMSE of 2891.57** indicates that, on average, your model's predictions are off by about 2891.57 units from the actual values.
**R² of 66.82%** shows that your model explains 66.82% of the variance in the target variable, which suggests a moderately strong model, but there may still be room for improvement.
The **mean value of 9263.97** provides a reference to compare the RMSE and understand the scale of errors relative to the overall dataset.
